File: floors/fortran/department_floor.f90
! FLOOR 24 - FORTRAN JURISDICTION
! Department Floor Implementation
!
! Domain: Scientific computation, Numerical analysis, HPC simulations
! Architectural Law: Numerical precision > convenience, Array bounds mandatory
! Security Doctrine: Array bounds checking, Numerical stability validation

program department_floor
    use iso_fortran_env, only: output_unit, error_unit, input_unit
    implicit none
    
    integer, parameter :: MAX_LINE = 8192
    integer, parameter :: MAX_AGENTS = 100
    integer, parameter :: MAX_TASKS = 100
    integer, parameter :: MAX_NAME = 256
    integer, parameter :: MAX_CODE = 16384
    
    character(len=MAX_LINE) :: line
    integer :: ios, request_id
    
    ! Initialize and run main loop
    call process_requests()
    
contains

    ! Main request processing loop
    subroutine process_requests()
        implicit none
        character(len=MAX_LINE) :: input_line
        integer :: io_status
        
        do
            ! Read from stdin
            read(input_unit, '(A)', iostat=io_status) input_line
            if (io_status /= 0) exit
            
            ! Process the request
            call handle_request(trim(adjustl(input_line)))
        end do
    end subroutine process_requests
    
    ! Handle a single JSON-RPC request
    subroutine handle_request(json_str)
        implicit none
        character(len=*), intent(in) :: json_str
        character(len=64) :: method
        integer :: req_id
        
        if (len_trim(json_str) == 0) return
        
        ! Extract method and id
        call extract_json_field(json_str, 'method', method)
        req_id = extract_json_int(json_str, 'id')
        
        ! Route to appropriate handler
        select case (trim(method))
            case ('initialize')
                call handle_initialize(req_id)
            case ('get_floor_info')
                call handle_get_floor_info(req_id)
            case ('add_agent')
                call handle_add_agent(json_str, req_id)
            case ('create_task')
                call handle_create_task(json_str, req_id)
            case ('analyze_code')
                call handle_analyze_code(json_str, req_id)
            case ('list_agents')
                call handle_list_agents(req_id)
            case ('list_tasks')
                call handle_list_tasks(req_id)
            case ('shutdown')
                call handle_shutdown(req_id)
                stop
            case default
                call send_error(req_id, -32601, 'Method not found')
        end select
    end subroutine handle_request
    
    ! Initialize response
    subroutine handle_initialize(req_id)
        implicit none
        integer, intent(in) :: req_id
        
        write(output_unit, '(A)', advance='no') '{"jsonrpc":"2.0","id":'
        call write_integer(req_id)
        write(output_unit, '(A)') ',"result":{"status":"initialized",' // &
            '"floor":24,"language":"fortran",' // &
            '"capabilities":["scientific_compute","numerical_analysis","array_operations"]}}'
        flush(output_unit)
    end subroutine handle_initialize
    
    ! Get floor information
    subroutine handle_get_floor_info(req_id)
        implicit none
        integer, intent(in) :: req_id
        
        write(output_unit, '(A)', advance='no') '{"jsonrpc":"2.0","id":'
        call write_integer(req_id)
        write(output_unit, '(A)') ',"result":{' // &
            '"floor_number":24,' // &
            '"language":"fortran",' // &
            '"domain":"Scientific computation, Numerical analysis, HPC simulations",' // &
            '"offices":["Architecture Office","Implementation Office",' // &
            '"Review Office","Test Office","Security Office","Manager Office"],' // &
            '"agents":[' // &
            '{"agent_id":"fortran-service-1","name":"Numerical Service Agent",' // &
            '"role":"Service","capabilities":["precision_control","rounding_analysis"]},' // &
            '{"agent_id":"fortran-data-1","name":"Array Data Agent",' // &
            '"role":"Data","capabilities":["array_validation","bounds_checking"]},' // &
            '{"agent_id":"fortran-ops-1","name":"Operations Agent",' // &
            '"role":"Operations","capabilities":["matrix_operations","numerical_methods"]}' // &
            '],' // &
            '"architectural_laws":[' // &
            '"Numerical precision over convenience",' // &
            '"Mandatory array bounds checking",' // &
            '"Explicit array dimensions",' // &
            '"Column-major ordering"' // &
            '],' // &
            '"security_doctrine":[' // &
            '"Array bounds validation",' // &
            '"Numerical stability checks",' // &
            '"Overflow detection",' // &
            '"Precision loss warnings"' // &
            ']}}'
        flush(output_unit)
    end subroutine handle_get_floor_info
    
    ! Add agent handler
    subroutine handle_add_agent(json_str, req_id)
        implicit none
        character(len=*), intent(in) :: json_str
        integer, intent(in) :: req_id
        character(len=64) :: agent_id, name, role
        
        call extract_json_field(json_str, 'agent_id', agent_id)
        call extract_json_field(json_str, 'name', name)
        call extract_json_field(json_str, 'role', role)
        
        write(output_unit, '(A)', advance='no') '{"jsonrpc":"2.0","id":'
        call write_integer(req_id)
        write(output_unit, '(A)', advance='no') ',"result":{"status":"success","agent":{"agent_id":"'
        write(output_unit, '(A)', advance='no') trim(agent_id)
        write(output_unit, '(A)', advance='no') '","name":"'
        write(output_unit, '(A)', advance='no') trim(name)
        write(output_unit, '(A)', advance='no') '","role":"'
        write(output_unit, '(A)', advance='no') trim(role)
        write(output_unit, '(A)') '"}}}'
        flush(output_unit)
    end subroutine handle_add_agent
    
    ! Create task handler
    subroutine handle_create_task(json_str, req_id)
        implicit none
        character(len=*), intent(in) :: json_str
        integer, intent(in) :: req_id
        character(len=64) :: task_id, title
        
        call extract_json_field(json_str, 'task_id', task_id)
        call extract_json_field(json_str, 'title', title)
        
        write(output_unit, '(A)', advance='no') '{"jsonrpc":"2.0","id":'
        call write_integer(req_id)
        write(output_unit, '(A)', advance='no') ',"result":{"status":"success","task":{"task_id":"'
        write(output_unit, '(A)', advance='no') trim(task_id)
        write(output_unit, '(A)', advance='no') '","title":"'
        write(output_unit, '(A)', advance='no') trim(title)
        write(output_unit, '(A)') '","status":"pending"}}}'
        flush(output_unit)
    end subroutine handle_create_task
    
    ! Analyze code - Scientific Computation Agent
    subroutine handle_analyze_code(json_str, req_id)
        implicit none
        character(len=*), intent(in) :: json_str
        integer, intent(in) :: req_id
        character(len=MAX_CODE) :: code
        integer :: lines, subroutines, arrays, do_loops
        
        call extract_json_field(json_str, 'code', code)
        
        ! Analyze code (Service Agent)
        call analyze_fortran_code(code, lines, subroutines, arrays, do_loops)
        
        write(output_unit, '(A)', advance='no') '{"jsonrpc":"2.0","id":'
        call write_integer(req_id)
        write(output_unit, '(A)', advance='no') ',"result":{"language":"fortran","lines":'
        call write_integer(lines)
        write(output_unit, '(A)', advance='no') ',"subroutines":'
        call write_integer(subroutines)
        write(output_unit, '(A)', advance='no') ',"arrays":'
        call write_integer(arrays)
        write(output_unit, '(A)', advance='no') ',"do_loops":'
        call write_integer(do_loops)
        write(output_unit, '(A)') ',"agent":"Numerical Service Agent"}}'
        flush(output_unit)
    end subroutine handle_analyze_code
    
    ! List agents
    subroutine handle_list_agents(req_id)
        implicit none
        integer, intent(in) :: req_id
        
        write(output_unit, '(A)', advance='no') '{"jsonrpc":"2.0","id":'
        call write_integer(req_id)
        write(output_unit, '(A)') ',"result":{"agents":[' // &
            '{"agent_id":"fortran-service-1","name":"Numerical Service Agent","role":"Service"},' // &
            '{"agent_id":"fortran-data-1","name":"Array Data Agent","role":"Data"},' // &
            '{"agent_id":"fortran-ops-1","name":"Operations Agent","role":"Operations"}' // &
            ']}}'
        flush(output_unit)
    end subroutine handle_list_agents
    
    ! List tasks
    subroutine handle_list_tasks(req_id)
        implicit none
        integer, intent(in) :: req_id
        
        write(output_unit, '(A)', advance='no') '{"jsonrpc":"2.0","id":'
        call write_integer(req_id)
        write(output_unit, '(A)') ',"result":{"tasks":[]}}'
        flush(output_unit)
    end subroutine handle_list_tasks
    
    ! Shutdown handler
    subroutine handle_shutdown(req_id)
        implicit none
        integer, intent(in) :: req_id
        
        write(output_unit, '(A)', advance='no') '{"jsonrpc":"2.0","id":'
        call write_integer(req_id)
        write(output_unit, '(A)') ',"result":{"status":"shutdown"}}'
        flush(output_unit)
    end subroutine handle_shutdown
    
    ! Error response
    subroutine send_error(req_id, error_code, error_msg)
        implicit none
        integer, intent(in) :: req_id, error_code
        character(len=*), intent(in) :: error_msg
        
        write(output_unit, '(A)', advance='no') '{"jsonrpc":"2.0","id":'
        call write_integer(req_id)
        write(output_unit, '(A)', advance='no') ',"error":{"code":'
        call write_integer(error_code)
        write(output_unit, '(A)', advance='no') ',"message":"'
        write(output_unit, '(A)', advance='no') trim(error_msg)
        write(output_unit, '(A)') '"}}'
        flush(output_unit)
    end subroutine send_error
    
    ! SERVICE AGENT: Analyze Fortran code for scientific computing patterns
    subroutine analyze_fortran_code(code, lines, subroutines, arrays, do_loops)
        implicit none
        character(len=*), intent(in) :: code
        integer, intent(out) :: lines, subroutines, arrays, do_loops
        integer :: i, code_len
        logical :: in_comment
        
        lines = 0
        subroutines = 0
        arrays = 0
        do_loops = 0
        code_len = len_trim(code)
        in_comment = .false.
        
        ! Count lines
        do i = 1, code_len
            if (code(i:i) == char(10)) lines = lines + 1
        end do
        if (lines == 0 .and. code_len > 0) lines = 1
        
        ! Count subroutines and functions
        subroutines = count_substring(code, 'subroutine') + &
                     count_substring(code, 'function')
        
        ! Count array declarations (dimension keyword)
        arrays = count_substring(code, 'dimension') + &
                count_substring(code, '::')
        
        ! Count DO loops
        do_loops = count_substring(code, 'do ') + &
                  count_substring(code, 'DO ')
    end subroutine analyze_fortran_code
    
    ! DATA/MODEL AGENT: Validate array bounds
    subroutine validate_array_bounds(array_size, index, is_valid)
        implicit none
        integer, intent(in) :: array_size, index
        logical, intent(out) :: is_valid
        
        is_valid = .false.
        if (index >= 1 .and. index <= array_size) then
            is_valid = .true.
        end if
    end subroutine validate_array_bounds
    
    ! OPERATIONS AGENT: Perform matrix operation with bounds checking
    subroutine safe_matrix_multiply(a, b, c, n, m, p, status)
        implicit none
        integer, intent(in) :: n, m, p
        real(8), intent(in) :: a(n,m), b(m,p)
        real(8), intent(out) :: c(n,p)
        integer, intent(out) :: status
        integer :: i, j, k
        
        status = 0
        
        ! Validate dimensions
        if (n <= 0 .or. m <= 0 .or. p <= 0) then
            status = -1
            return
        end if
        
        ! Perform multiplication with numerical stability
        c = 0.0d0
        do j = 1, p
            do i = 1, n
                do k = 1, m
                    c(i,j) = c(i,j) + a(i,k) * b(k,j)
                end do
            end do
        end do
    end subroutine safe_matrix_multiply
    
    ! Helper: Count substring occurrences
    function count_substring(text, pattern) result(count)
        implicit none
        character(len=*), intent(in) :: text, pattern
        integer :: count
        integer :: pos, text_len, pattern_len
        
        count = 0
        text_len = len_trim(text)
        pattern_len = len_trim(pattern)
        pos = 1
        
        do while (pos <= text_len - pattern_len + 1)
            if (text(pos:pos+pattern_len-1) == pattern) then
                count = count + 1
                pos = pos + pattern_len
            else
                pos = pos + 1
            end if
        end do
    end function count_substring
    
    ! Helper: Extract JSON field
    subroutine extract_json_field(json_str, key, value)
        implicit none
        character(len=*), intent(in) :: json_str, key
        character(len=*), intent(out) :: value
        character(len=512) :: search_str
        integer :: key_pos, start_pos, end_pos
        
        value = ''
        search_str = '"' // trim(key) // '"'
        key_pos = index(json_str, trim(search_str))
        
        if (key_pos == 0) return
        
        start_pos = index(json_str(key_pos:), ':')
        if (start_pos == 0) return
        start_pos = key_pos + start_pos
        
        ! Skip whitespace and opening quote
        do while (start_pos <= len(json_str))
            if (json_str(start_pos:start_pos) == '"') then
                start_pos = start_pos + 1
                exit
            end if
            start_pos = start_pos + 1
        end do
        
        ! Find closing quote
        end_pos = start_pos
        do while (end_pos <= len(json_str))
            if (json_str(end_pos:end_pos) == '"') exit
            end_pos = end_pos + 1
        end do
        
        if (end_pos > start_pos) then
            value = json_str(start_pos:end_pos-1)
        end if
    end subroutine extract_json_field
    
    ! Helper: Extract JSON integer
    function extract_json_int(json_str, key) result(value)
        implicit none
        character(len=*), intent(in) :: json_str, key
        integer :: value
        character(len=64) :: value_str
        integer :: ios
        
        call extract_json_field(json_str, key, value_str)
        read(value_str, *, iostat=ios) value
        if (ios /= 0) value = 0
    end function extract_json_int
    
    ! Helper: Write integer to output
    subroutine write_integer(n)
        implicit none
        integer, intent(in) :: n
        character(len=32) :: str
        
        write(str, '(I0)') n
        write(output_unit, '(A)', advance='no') trim(str)
    end subroutine write_integer

end program department_floor

File: floors/matlab/department_floor.m
% FLOOR 25 - MATLAB/OCTAVE JURISDICTION
% Department Floor Implementation
%
% Domain: Numerical modeling, Matrix operations, Signal processing, Visualization
% Architectural Law: Vectorization > loops, Matrix dimensions explicit
% Security Doctrine: Dimension validation, NaN/Inf checking, Index bounds

function department_floor()
    % Main entry point - read JSON-RPC from stdin, write to stdout
    
    while true
        try
            line = input('', 's');
            if isempty(line)
                break;
            end
            handle_request(line);
        catch ME
            % End of input or error
            break;
        end
    end
end

function handle_request(json_str)
    % Parse and route JSON-RPC request
    
    try
        req = parse_json(json_str);
        
        if ~isfield(req, 'method') || ~isfield(req, 'id')
            send_error(0, -32600, 'Invalid Request');
            return;
        end
        
        method = req.method;
        req_id = req.id;
        
        % Route to appropriate handler
        switch method
            case 'initialize'
                handle_initialize(req_id);
            case 'get_floor_info'
                handle_get_floor_info(req_id);
            case 'add_agent'
                handle_add_agent(req, req_id);
            case 'create_task'
                handle_create_task(req, req_id);
            case 'analyze_code'
                handle_analyze_code(req, req_id);
            case 'list_agents'
                handle_list_agents(req_id);
            case 'list_tasks'
                handle_list_tasks(req_id);
            case 'shutdown'
                handle_shutdown(req_id);
                return;
            otherwise
                send_error(req_id, -32601, 'Method not found');
        end
    catch ME
        send_error(0, -32700, 'Parse error');
    end
end

function handle_initialize(req_id)
    % Initialize the floor
    
    result = struct();
    result.status = 'initialized';
    result.floor = 25;
    result.language = 'matlab';
    result.capabilities = {'matrix_operations', 'numerical_modeling', 'signal_processing'};
    
    send_response(req_id, result);
end

function handle_get_floor_info(req_id)
    % Return comprehensive floor information
    
    result = struct();
    result.floor_number = 25;
    result.language = 'matlab';
    result.domain = 'Numerical modeling, Matrix operations, Signal processing, Visualization';
    
    result.offices = {'Architecture Office', 'Implementation Office', ...
                      'Review Office', 'Test Office', ...
                      'Security Office', 'Manager Office'};
    
    % Define agents (Service, Data, Operations)
    agents = {};
    
    agent1 = struct();
    agent1.agent_id = 'matlab-service-1';
    agent1.name = 'Matrix Service Agent';
    agent1.role = 'Service';
    agent1.capabilities = {'vectorization', 'matrix_analysis', 'performance_optimization'};
    agents{1} = agent1;
    
    agent2 = struct();
    agent2.agent_id = 'matlab-data-1';
    agent2.name = 'Data Model Agent';
    agent2.role = 'Data';
    agent2.capabilities = {'dimension_validation', 'type_checking', 'data_structure_management'};
    agents{2} = agent2;
    
    agent3 = struct();
    agent3.agent_id = 'matlab-ops-1';
    agent3.name = 'Operations Agent';
    agent3.role = 'Operations';
    agent3.capabilities = {'linear_algebra', 'signal_processing', 'numerical_methods'};
    agents{3} = agent3;
    
    result.agents = agents;
    
    result.architectural_laws = {'Vectorization over loops', ...
                                   'Explicit matrix dimensions', ...
                                   'Preallocate arrays', ...
                                   'Avoid dynamic resizing'};
    
    result.security_doctrine = {'Dimension validation before operations', ...
                                 'NaN and Inf checking', ...
                                 'Index bounds verification', ...
                                 'Condition number validation'};
    
    send_response(req_id, result);
end

function handle_add_agent(req, req_id)
    % Add a new agent
    
    params = req.params;
    
    agent = struct();
    agent.agent_id = get_field(params, 'agent_id', 'unknown');
    agent.name = get_field(params, 'name', 'Unknown Agent');
    agent.role = get_field(params, 'role', 'General');
    
    result = struct();
    result.status = 'success';
    result.agent = agent;
    
    send_response(req_id, result);
end

function handle_create_task(req, req_id)
    % Create a new task
    
    params = req.params;
    
    task = struct();
    task.task_id = get_field(params, 'task_id', 'unknown');
    task.title = get_field(params, 'title', 'Untitled Task');
    task.status = 'pending';
    task.assigned_to = get_field(params, 'assigned_to', 'unassigned');
    
    result = struct();
    result.status = 'success';
    result.task = task;
    
    send_response(req_id, result);
end

function handle_analyze_code(req, req_id)
    % Analyze MATLAB/Octave code - SERVICE AGENT
    
    params = req.params;
    code = get_field(params, 'code', '');
    
    % Analyze code structure
    analysis = analyze_matlab_code(code);
    
    result = struct();
    result.language = 'matlab';
    result.lines = analysis.lines;
    result.functions = analysis.functions;
    result.matrices = analysis.matrices;
    result.vectorized_ops = analysis.vectorized_ops;
    result.agent = 'Matrix Service Agent';
    
    send_response(req_id, result);
end

function handle_list_agents(req_id)
    % List all agents
    
    agents = {};
    
    agent1 = struct('agent_id', 'matlab-service-1', 'name', 'Matrix Service Agent', 'role', 'Service');
    agent2 = struct('agent_id', 'matlab-data-1', 'name', 'Data Model Agent', 'role', 'Data');
    agent3 = struct('agent_id', 'matlab-ops-1', 'name', 'Operations Agent', 'role', 'Operations');
    
    agents{1} = agent1;
    agents{2} = agent2;
    agents{3} = agent3;
    
    result = struct();
    result.agents = agents;
    
    send_response(req_id, result);
end

function handle_list_tasks(req_id)
    % List all tasks
    
    result = struct();
    result.tasks = {};
    
    send_response(req_id, result);
end

function handle_shutdown(req_id)
    % Shutdown the floor
    
    result = struct();
    result.status = 'shutdown';
    
    send_response(req_id, result);
end

% ============================================================================
% SERVICE AGENT: Code Analysis
% ============================================================================

function analysis = analyze_matlab_code(code)
    % Analyze MATLAB code structure and patterns
    
    analysis = struct();
    
    % Count lines
    lines = split(code, newline);
    analysis.lines = length(lines);
    
    % Count functions
    analysis.functions = count_pattern(code, 'function');
    
    % Count matrix operations
    analysis.matrices = count_pattern(code, '\[') + count_pattern(code, 'zeros') + ...
                        count_pattern(code, 'ones') + count_pattern(code, 'eye');
    
    % Count vectorized operations
    analysis.vectorized_ops = count_pattern(code, '.*') + count_pattern(code, './') + ...
                              count_pattern(code, '.^') + count_pattern(code, '''');
end

function count = count_pattern(text, pattern)
    % Count occurrences of pattern in text
    
    count = 0;
    if isempty(text) || isempty(pattern)
        return;
    end
    
    matches = strfind(lower(text), lower(pattern));
    count = length(matches);
end

% ============================================================================
% DATA/MODEL AGENT: Dimension Validation
% ============================================================================

function is_valid = validate_matrix_dimensions(A, B, operation)
    % Validate matrix dimensions for operations
    
    is_valid = false;
    
    if isempty(A) || isempty(B)
        return;
    end
    
    [m1, n1] = size(A);
    [m2, n2] = size(B);
    
    switch operation
        case 'multiply'
            is_valid = (n1 == m2);
        case 'add'
            is_valid = (m1 == m2) && (n1 == n2);
        case 'subtract'
            is_valid = (m1 == m2) && (n1 == n2);
        otherwise
            is_valid = false;
    end
end

function is_valid = validate_vector_index(vec, idx)
    % Validate vector index bounds
    
    is_valid = false;
    
    if isempty(vec) || idx < 1
        return;
    end
    
    n = length(vec);
    is_valid = (idx <= n);
end

% ============================================================================
% OPERATIONS AGENT: Matrix Operations
% ============================================================================

function [C, status] = safe_matrix_multiply(A, B)
    % Safely multiply matrices with dimension validation
    
    status = 0;
    C = [];
    
    % Validate inputs
    if ~isnumeric(A) || ~isnumeric(B)
        status = -1;
        return;
    end
    
    % Check dimensions
    if ~validate_matrix_dimensions(A, B, 'multiply')
        status = -2;
        return;
    end
    
    % Perform multiplication
    try
        C = A * B;
        
        % Check for NaN or Inf
        if any(isnan(C(:))) || any(isinf(C(:)))
            status = -3;
        end
    catch
        status = -4;
    end
end

function [v, status] = safe_vector_access(vec, idx)
    % Safely access vector element with bounds checking
    
    status = 0;
    v = NaN;
    
    if ~validate_vector_index(vec, idx)
        status = -1;
        return;
    end
    
    v = vec(idx);
end

% ============================================================================
% JSON Utilities
% ============================================================================

function req = parse_json(json_str)
    % Simple JSON parser for basic structures
    
    % Try using built-in JSON decoder if available (MATLAB R2016b+)
    if exist('jsondecode', 'file')
        req = jsondecode(json_str);
        return;
    end
    
    % Fallback: manual parsing for basic JSON
    req = struct();
    
    % Extract method
    method_match = regexp(json_str, '"method"\s*:\s*"([^"]+)"', 'tokens');
    if ~isempty(method_match)
        req.method = method_match{1}{1};
    end
    
    % Extract id
    id_match = regexp(json_str, '"id"\s*:\s*(\d+)', 'tokens');
    if ~isempty(id_match)
        req.id = str2double(id_match{1}{1});
    else
        req.id = 0;
    end
    
    % Extract params if present
    if contains(json_str, '"params"')
        req.params = struct();
        
        % Extract simple string params
        param_matches = regexp(json_str, '"(\w+)"\s*:\s*"([^"]*)"', 'tokens');
        for i = 1:length(param_matches)
            key = param_matches{i}{1};
            value = param_matches{i}{2};
            if ~strcmp(key, 'method') && ~strcmp(key, 'jsonrpc')
                req.params.(key) = value;
            end
        end
    end
end

function send_response(req_id, result)
    % Send JSON-RPC response
    
    % Try using built-in JSON encoder if available
    if exist('jsonencode', 'file')
        response = struct();
        response.jsonrpc = '2.0';
        response.id = req_id;
        response.result = result;
        fprintf('%s\n', jsonencode(response));
    else
        % Fallback: manual JSON construction
        fprintf('{"jsonrpc":"2.0","id":%d,"result":', req_id);
        print_struct(result);
        fprintf('}\n');
    end
    
    % Flush output
    if exist('OCTAVE_VERSION', 'builtin')
        fflush(stdout);
    else
        drawnow;
    end
end

function send_error(req_id, error_code, error_msg)
    % Send JSON-RPC error response
    
    fprintf('{"jsonrpc":"2.0","id":%d,"error":{"code":%d,"message":"%s"}}\n', ...
            req_id, error_code, error_msg);
    
    if exist('OCTAVE_VERSION', 'builtin')
        fflush(stdout);
    else
        drawnow;
    end
end

function print_struct(s)
    % Print struct as JSON (simple implementation)
    
    fprintf('{');
    fields = fieldnames(s);
    for i = 1:length(fields)
        if i > 1
            fprintf(',');
        end
        fprintf('"%s":', fields{i});
        
        value = s.(fields{i});
        if ischar(value)
            fprintf('"%s"', value);
        elseif isnumeric(value) && isscalar(value)
            fprintf('%g', value);
        elseif iscell(value)
            fprintf('[');
            for j = 1:length(value)
                if j > 1
                    fprintf(',');
                end
                if isstruct(value{j})
                    print_struct(value{j});
                elseif ischar(value{j})
                    fprintf('"%s"', value{j});
                end
            end
            fprintf(']');
        elseif isstruct(value)
            print_struct(value);
        end
    end
    fprintf('}');
end

function value = get_field(s, field, default)
    % Safely get field from struct with default
    
    if isfield(s, field)
        value = s.(field);
    else
        value = default;
    end
end
